### PR TITLE
#82: Extend key file support

### PR DESCRIPTION
**Pull request by tobischo**
Extends support of parsing key files for KDBX files

### CODE REVIEW

File: credentials.go [hunk 1]
```diff
@@ -145,24 +147,73 @@ func ParseKeyFile(location string) ([]byte, error) {
...
+	err := xml.Unmarshal(data, &keyFileData)
+	if err == nil {
+		keyFileData.Key.Data = whiteSpacePattern.ReplaceAll(keyFileData.Key.Data, []byte(``))
+
+		switch keyFileData.Meta.Version {
+		case "1.00", "1.0":
+			decodedKey := make([]byte, base64.StdEncoding.DecodedLen(len(keyFileData.Key.Data)))
+			_, err := base64.StdEncoding.Decode(decodedKey, keyFileData.Key.Data)
+			return decodedKey, err
+		case "2.0":
```
Comment (Footmau) [inline credentials.go:183]: Can it be also declared as "2.00"?

File: credentials.go [hunk 2]
```diff
@@ -145,24 +147,73 @@ func ParseKeyFile(location string) ([]byte, error) {
...
 		return data, nil
 	}
 
-	// Slice necessary due to padding at the end of the hash
-	return data[:32], nil
+	if len(data) == 64 {
+		decodedHex, err := hex.DecodeString(string(data))
+		if err == nil {
+			return decodedHex, nil
+		}
```
Comment (Footmau) [inline credentials.go:212]: Should we return error or it is fine to go farther?

File: credentials.go [hunk 3]
```diff
@@ -145,24 +147,73 @@ func ParseKeyFile(location string) ([]byte, error) {
...
-		base := keyDataPattern.FindSubmatch(data)[1]
-		data = make([]byte, base64.StdEncoding.DecodedLen(len(base)))
-		if _, err := base64.StdEncoding.Decode(data, base); err != nil {
-			return nil, err
+	keyFileData := xmlKeyFileData{}
+	err := xml.Unmarshal(data, &keyFileData)
+	if err == nil {
+		keyFileData.Key.Data = whiteSpacePattern.ReplaceAll(keyFileData.Key.Data, []byte(``))
+
+		switch keyFileData.Meta.Version {
```
Comment (Footmau) [inline credentials.go:178]: Probably it is better to define separate decode funcs for every case. This one looks a bit overloaded.

File: credentials.go (hunk 1 above)
Comment (tobischo) [reply credentials.go:183]: According to the comment in the original source it was only `1.00` for some versions of the code, and the other would be `1.0` and `2.0`

File: credentials.go (hunk 3 above)
Comment (tobischo) [reply credentials.go:178]: Yup, I only threw this together as a first draft

File: credentials.go (hunk 2 above)
Comment (tobischo) [reply credentials.go:212]: The idea was (based on the original keepass source) if it cannot be parsed as a hex string, then it needs to be hashed, even if it is (by accident) 64 chars long

File: credentials.go (hunk 3 above)
Comment (tobischo) [reply credentials.go]: I split it out into smaller chunks

I still need to get some test data with different types of keys in order to finish this

Comment (Footmau) [reply credentials.go]: Do you expect the test data from me? I can prepare and upload it here. Alternatively test data can be generated randomly (with optional xml wrapper), but it that way we can not make sure if it works with original KeePass app.

**Review by Footmau — APPROVED** [review]

File: credentials.go (hunk 3 above)
Comment (tobischo) [reply credentials.go]: I generated test data by myself, however that is based on my memory of what should work.

I will however not get towards being able to generate database files + corresponding key files in the next 2 days to ensure that that part works

Comment (Footmau) [reply credentials.go]: Clear. I'll check with manual tests with databases and various key files then. Will report in 2 days as well.

**Pull request merged**